ETL use graph modified times to only re-index datasets that have changed

callum-oakley opened the pull request:
Changes the ETL pipeline so that only observations for datasets that have changed are re-indexed. Achieved by:

- Indexing modified times for each observations graph
- Including the containing graph in the observations index so we can look up observations by graph
- Tracking which graphs have been added/removed/modified when updating the graph index
- Removing observations for graphs which have been removed or modified
- Adding observations for graphs which have been added or modified

Other things that had to change to make this happen:

- Pipeline functions return the number of documents that they indexed, mainly for testing purposes
- The individual pipeline functions are now responsible for deleting/creating their own indices -- in particular the observation and graph pipelines now depend on the previous state so we can't just wipe them all between runs
- The integrant system now includes `:ook.concerns.elastic/conn` instead of `:ook.concerns.elastic/endpoint` because all we ever did with the endpoint was wrap it in a connection
- Currently `idp-beta.edn` is pointing at `https://drafter-bump-drafter.publishmydata.com`, which is running a version of drafter with the modified times graph. Some of the tests have changed because the data is slightly different, but I updated the tests against `main`, so I'm confident they are correct (bugs in `main` notwithstanding)

Open questions/gotchas:

- `graph->modified` reads the entire graph index, but assumes it contains at most 10000 documents (since that's the size limit on the query). I think this is a reasonable assumption for the time-being? But if not we'll have to do some paging or store the `graph->modified` map elsewhere (just writing it to disk as a blob of JSON or EDN would probably be sufficient)
- We're using the modified times to determine if a graph has changed, but technically there's a race condition here if a graph changes twice within a millisecond (which appears to be [likely to actually happen](https://github.com/Swirrl/drafter/issues/455#issuecomment-808085519)). The solution to this would be to use versions instead, which drafter now uses internally but which are not currently exposed by the public modified time graph. It would be a fairly small change to expose them and then update this work to make use of them (in a follow up PR?).
- That being said... maybe we don't really care about the above at all because I don't think we can guarantee the elastic search indices are consistent anyway without taking a snapshot of the upstream data.

closes #17

callum-oakley committed:
bring bin/repl in line with bin/cider

callum-oakley committed:
return number of subjects loaded from pipeline functions

callum-oakley committed:
populate a graph index with modified times

callum-oakley committed:
point back to graph in observation index

callum-oakley committed:
full pipline working -- needs cleaning up

callum-oakley committed:
pipeline functions responsible for deleting/creating their own indices

callum-oakley committed:
remove a level of nesting from pipeline functions

callum-oakley committed:
initiate an elasticsearch connection with integrant so we don't have to call connect every time we want one

callum-oakley committed:
shuffle dependencies so that etl depends on index rather than the inverse

callum-oakley committed:
point to staging drafter with bumped drafter and update cassettes and tests. this change was made against main and then cherry picked, so the updates to the tests should be correct (bugs in main notwithstanding)

callum-oakley committed:
remove local.edn committed in error

callum-oakley committed:
generate cassettes

callum-oakley committed:
formatting

callum-oakley committed:
use with-redefs instead of intern for faking in test

Robsteranium reviewed (commented):
Hi, Sorry I left a review [in a comment](https://github.com/Swirrl/ook/pull/115#issuecomment-1047656926) and forgot to use this button!

callum-oakley committed:
point tests back at idp-beta drafter

callum-oakley committed:
comment

callum-oakley committed:
ook.index/data -> ook.etl/load

Robsteranium committed:
Have Circle build etl-use-graph-versions for testing on deployed server

Robsteranium committed:
Rename observation-graph.sparql -> graph-select.sparql for consistency with other indexes

Robsteranium committed:
Create observations index in pipeline if doesn't already exist

Robsteranium committed:
Reduce page size on code pipeline to 20k subjects

Robsteranium committed:
Reduce page size on code pipeline to 2k subjects

Robsteranium committed:
Reduce page size on code pipeline to 200 subjects

Robsteranium committed:
Set page size on code pipeline to 500 subjects (200 not compatible with existing cassettes)

Robsteranium committed:
Drop select-page-size to 20,000

Robsteranium commented:
We might want to resolve #120 as part of this PR.

Robsteranium committed:
Drop select-page-size to 10,000

Robsteranium committed:
Dump construct query to tempfile on retry during ETL

Robsteranium committed:
Log full construct query with interpolated values on retry

callum-oakley committed:
delete graph and observation indices on error

callum-oakley committed:
properly parse csv query results

callum-oakley committed:
buildme

callum-oakley committed:
cassettes refreshed with up to date data, tests all passing bar get-datasets-for-facets

Robsteranium committed:
Merge pull request #124 from Swirrl/parse-csv-query-results

properly parse csv query results

Robsteranium committed:
Merge pull request #123 from Swirrl/etl-use-graph-versions-transactional

delete graph and observation indices on error

Robsteranium commented:
I've confirmed that this works with additions but haven't be able to confirm with a deletion yet (pending examples from ONS). I suggest we also aim to resolve #120 before merging.

Robsteranium committed:
Merge pull request #125 from Swirrl/refresh-cassettes

Refresh cassettes.
Resolves #119

Robsteranium committed:
Use io/reader for load-data sparql targets

We were otherwise seeing:

> No implementation of method: :read-csv-from of protocol: #'clojure.data.csv/Read-CSV-From
> found for class: clj_http.core.proxy$java.io.FilterInputStream$ff19274a

Robsteranium committed:
Add config for :ook.etl/select-small-page-size for code and codes-used pipelines

Robsteranium committed:
Have CI build on etl-use-graph-versions branch

Robsteranium commented:
I can confirm that deletes work. I'm going to merge this now as it's been open for 6 months! We can resolve #120 separately.
